### PR TITLE
make JWT claim more resilient

### PR DESCRIPTION
**Pull request by ijlyttle**
I want to test a few things, so I'll play around in this PR - it will likely not come to anything

### CODE REVIEW

**Comment by ijlyttle** [comment]
I *think* this takes care of the bad-request problem to fix #166 .  I was having the problem again this morning, this seems to solve it.

I think this arises from the clock on a computer at Box not agreeing with the clock on the computer that makes the JWT claim.

Box lets you make a claim on how long your token assertion is valid for, it can be no more than 60 seconds into the future, as measured by Box.

Our default was to set the validity time to 45 seconds. If the clocks disagree such that Box thinks the assertion is no longer valid, voilà, bad request.

The proposed solution is to try a sequence of validity times: 30, 15, 45, 0, then 60 seconds. If there's a bad request, we issue a message then try with the next validity-time.

@jrosen48, @nathancday: would you try out:

```r
install_github("r-box/boxr", ref = "refactor-token")
```

to see if this takes care of the problem for you?

Thanks!

File: R/boxr_auth.R [hunk 1]
```diff
@@ -431,32 +431,67 @@ box_auth_service <- function(token_file = NULL, token_text = NULL) {
...
-  )
   
-  # sign claim with key
-  assertion <- jose::jwt_encode_sig(
-    claim, 
-    key,
-    header = list("kid" = config$boxAppSettings$appAuth$publicKeyID)
-  )
+  # wrap params in function, enable retry with different expiry times
+  params_dt <- function(dt = 30) {
```
Comment (nathancday) [inline R/boxr_auth.R:436]: The naming `dt` doesn't describe the object/idea to me. Would `exp_offset` or `exp_seconds` or something about the fact we are "padding" the system time (to generate the expiration time) be more helpful to a reader?

**Review by nathancday — CHANGES_REQUESTED** [review]
Still trying to wrap my brain around why this solves the API fails, but it's working all the time where straight `RETRY` wasn't. Great find and elegant code changes, just the one comment to discuss about descriptive names for `dt`

File: R/boxr_auth.R (hunk 1 above)
Comment (ijlyttle) [reply R/boxr_auth.R:436]: Let me see what I can do here.

**Comment by ijlyttle** [comment]
Let's say that my computer says it's 7:38:00, and the Box computer says it's 7:38:32 (how they could be a minute out is beyond me, and is frankly surprising).

On my first try I use `dt` of 30, so that I assert the claim is valid until 7:38:30. Box rejects the claim because, according to it, it's past the assertion deadline.

So I send another claim, using `dt` of 45, asserting it's valid until 7:38:47 (a couple of seconds have passed). At Box, it's now 7:38:34, so the claim is OK.

I imagine this is also the case where the Box clock is behind the computer's clock, and will not accept claims that are too far in the future.

Using only `RETRY`, I think we run up against the same clock offset - maybe it works sometimes because one of the clocks is reset (or we are asking another computer at Box).

File: R/boxr_auth.R (hunk 1 above)
Comment (ijlyttle) [reply R/boxr_auth.R]: I changed a few things around and changed the messages.

Thankfully, I was able to "reproduce" the error so I could confirm things work as I expect:

```
Failed JWT request: time offset was 0 seconds.
Retrying JWT request: time offset now -15 seconds.
Failed JWT request: time offset was -15 seconds.
Retrying JWT request: time offset now 15 seconds.
boxr: Authenticated using OAuth2 (JWT) as boxr-jwt-ijl (AutomationUser_868610_9SFx0Bq15D@boxdevedition.com, id: 9862360558)
```

I'll also add some words in the function doc and service-auth vignette - I'll do that in the docs PR

**Comment by ijlyttle** [comment]
@nathancday: I think (hope) this is ready to go, pending your acceptance and the Travis check

**Review by nathancday — APPROVED** [review]
Thanks for changing the names. Code looks good to go.

**Pull request merged**